bindings/py/tests/encoders/simhash_document_encoder_test.py

```diff
@@ -0,0 +1,282 @@
+  # of similar spelling will NOT affect the output in shared manner, but apart.
+  def testTokenSimilarityOff(self):
+    params = SimHashDocumentEncoderParameters()
+    params.size = 400
+    params.sparsity = 0.33
+    params.tokenSimilarity = False
+
+    encoder1 = SimHashDocumentEncoder(params)
+    encoder2 = SimHashDocumentEncoder(params)
+    encoder3 = SimHashDocumentEncoder(params)
+    encoder4 = SimHashDocumentEncoder(params)
+    output1 = SDR(params.size)
+    output2 = SDR(params.size)
+    output3 = SDR(params.size)
+    output4 = SDR(params.size)
+    encoder1.encode(testDoc1, output1)
+    encoder2.encode(testDoc2, output2)
+    encoder3.encode(testDoc3, output3)
+    encoder4.encode(testDoc4, output4)
+
+    assert(output1.getOverlap(output2) > output2.getOverlap(output3))
+    assert(output2.getOverlap(output3) > output3.getOverlap(output4))
+    assert(output3.getOverlap(output4) > output1.getOverlap(output3))
+
+  # Test encoding with weighted tokens. Make sure output changes accordingly.
+  def testTokenWeightMap(self):
+    params = SimHashDocumentEncoderParameters()
+    params.size = 400
+    params.sparsity = 0.33
+
+    encoder1 = SimHashDocumentEncoder(params)
+    encoder2 = SimHashDocumentEncoder(params)
+    encoder3 = SimHashDocumentEncoder(params)
+    output1 = SDR(params.size)
+    output2 = SDR(params.size)
+    output3 = SDR(params.size)
+    encoder1.encode(testDocMap1, output1)
+    encoder2.encode(testDocMap2, output2)
+    encoder3.encode(testDocMap3, output3)
+
+    assert(output1.getOverlap(output3) > output1.getOverlap(output2))
+    assert(output1.getOverlap(output2) > output2.getOverlap(output3))
+
+  # Test encoding unicode text with 'tokenSimilarity' on
+  def testUnicodeSimilarityOn(self):
+    params = SimHashDocumentEncoderParameters()
+    params.size = 400
+    params.sparsity = 0.33
+    params.tokenSimilarity = True
+
+    encoder1 = SimHashDocumentEncoder(params)
+    encoder2 = SimHashDocumentEncoder(params)
+    output1 = SDR(params.size)
+    output2 = SDR(params.size)
+    encoder1.encode(testDocUni1, output1)
+    encoder2.encode(testDocUni2, output2)
+
+    assert(output1.getOverlap(output2) > 65)
+
+  # Test encoding unicode text with 'tokenSimilarity' Off
+  def testUnicodeSimilarityOff(self):
+    params = SimHashDocumentEncoderParameters()
+    params.size = 400
+    params.sparsity = 0.33
+    params.tokenSimilarity = False
+
+    encoder1 = SimHashDocumentEncoder(params)
+    encoder2 = SimHashDocumentEncoder(params)
+    output1 = SDR(params.size)
+    output2 = SDR(params.size)
+    encoder1.encode(testDocUni1, output1)
+    encoder2.encode(testDocUni2, output2)
+
+    assert(output1.getOverlap(output2) < 65)
+
+  # Test serialization and deserialization
+  @pytest.mark.skip(reason="Known issue: https://github.com/htm-community/htm.core/issues/160")
+  def testPickle(self):
+    assert(False)  # @TODO: Serialization Unimplemented
```

Comment:
as this is a c++ encoder, it should implement c++ Serializable, and you could test that.

Reply:
uh, sorry, I just followed the other 2 encoders here, I don't know details.
Would you mind a bit of a code example?  thanks :)

Reply:
I've made some progress here, but it's taken hours, and I'm stuck now.  Is it cool if we move this over to PR #608?  All 3 encoders (this, RDSE, Scalar) will need a bit of synchronized work. I'll submit a new PR against https://github.com/htm-community/htm.core/tree/serialization_rdse with my work so far, and details.  The string serialization works great. The pickle is still failing on deserialize, I've narrowed it down to where the problem is occurring, but we'll have to dig deeper into some internals to trace what's going on. It's weird, I have no idea so far.

Reply:
Thanks for looking into the serialization problem! Definitely ok to have the test prepared as is now, and resolve the bug in #608